imessage-database/src/util/plist.rs

```diff
@@ -0,0 +1,240 @@
+                    ))?;
+                // Die here if the data is invalid
+                if keys.len() != values.len() {
+                    return Err(PlistParseError::InvalidDictionarySize(
+                        keys.len(),
+                        values.len(),
+                    ));
+                }
+
+                for idx in 0..keys.len() {
+                    let key_index = keys
+                        .get(idx)
+                        .ok_or(PlistParseError::NoValueAtIndex(idx))?
+                        .as_uid()
+                        .ok_or_else(|| PlistParseError::InvalidTypeIndex(idx, "uid".to_string()))?
+                        .get() as usize;
+                    let value_index = values
+                        .get(idx)
+                        .ok_or(PlistParseError::NoValueAtIndex(idx))?
+                        .as_uid()
+                        .ok_or_else(|| PlistParseError::InvalidTypeIndex(idx, "uid".to_string()))?
+                        .get() as usize;
+
+                    let key = objects
+                        .get(key_index)
+                        .ok_or(PlistParseError::NoValueAtIndex(key_index))?
+                        .as_string()
+                        .ok_or_else(|| PlistParseError::InvalidTypeIndex(
+                            key_index,
+                            "string".to_string(),
+                        ))?;
+
+                    dictionary.insert(
+                        key.to_string(),
+                        follow_uid(objects, value_index, Some(key.to_string()), None)?,
+                    );
+                }
+            }
+            // Handle a normal `{key: value}` style dictionary
+            else {
+                for (key, val) in dict {
+                    // Handle skips
+                    if key == "$class" {
+                        continue;
+                    }
+                    // If the value is a pointer, follow it
+                    if let Some(idx) = val.as_uid() {
+                        dictionary.insert(
+                            key.to_owned(),
+                            follow_uid(objects, idx.get() as usize, Some(key.to_string()), None)?,
+                        );
+                    }
+                    // If the value is not a pointer, try and follow the data itself
+                    else if let Some(p) = &parent {
+                        dictionary.insert(
+                            p.to_owned(),
+                            follow_uid(objects, root, Some(p.to_string()), Some(val))?,
+                        );
+                    }
+                }
+            }
+            Ok(plist::Value::Dictionary(dictionary))
+        }
+        Value::Uid(uid) => {
+            follow_uid(objects, uid.get() as usize, None, None)
+        }
+        Value::Boolean(_) | Value::Integer(_) | Value::String(_) => Ok(item.to_owned()),
+        _ => Ok(item.to_owned()),
```

Comment:
Can this just be a catch-all? Or do you prefer to enumerate these base types for clarity?